rclcpp/src/rclcpp/node_interfaces/node_graph.cpp

```diff
@@ -129,30 +130,49 @@ NodeGraph::get_service_names_and_types() const
 
 std::vector<std::string>
 NodeGraph::get_node_names() const
+{
+  std::vector<std::string> nodes;
+  auto names_and_namespaces = get_node_names_and_namespaces();
+
+  for (const auto & it : names_and_namespaces) {
+    nodes.push_back(it.first);
+  }
+  return nodes;
+}
+
+std::vector<std::pair<std::string, std::string>>
+NodeGraph::get_node_names_and_namespaces() const
 {
   rcutils_string_array_t node_names_c =
     rcutils_get_zero_initialized_string_array();
+  rcutils_string_array_t node_namespaces_c =
+    rcutils_get_zero_initialized_string_array();
 
   auto allocator = rcl_get_default_allocator();
   auto ret = rcl_get_node_names(
     node_base_->get_rcl_node_handle(),
     allocator,
-    &node_names_c);
+    &node_names_c,
+    &node_namespaces_c);
   if (ret != RCL_RET_OK) {
     auto error_msg = std::string("failed to get node names: ") + rcl_get_error_string_safe();
     rcl_reset_error();
     if (rcutils_string_array_fini(&node_names_c) != RCUTILS_RET_OK) {
       error_msg += std::string(", failed also to cleanup node names, leaking memory: ") +
         rcl_get_error_string_safe();
     }
+    if (rcutils_string_array_fini(&node_namespaces_c) != RCUTILS_RET_OK) {
+      error_msg += std::string(", failed also to cleanup node namespaces, leaking memory: ") +
+        rcl_get_error_string_safe();
```

Comment:
This wasn't being done in the existing code either, but you should reset the error with `rcl_reset_error()`.